### PR TITLE
refactor: don't hardcode the transaction version

### PR DESCRIPTION
**Pull request by panleone**
## Issue being fixed or feature implemented
Refactor extracted from #5860. Even if that PR should not need anymore this commit (since it has been found a better way to activate dip143), I think It's still a worthy refactor

## What was done?
  Introduce  `tx.IsSpecialTxVersion()` in place of `tx.nVersion == 3`,  `tx.nVersion >= 3` 
  and `tx.HasExtraPayloadField()` in place of  `tx.nVersion == 3 && tx.nType != TRANSACTION_NORMAL`

## Checklist:
  _Go over all the following points, and put an `x` in all the boxes that apply._
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/primitives/transaction.h [hunk 1]
```diff
@@ -304,6 +314,16 @@ struct CMutableTransaction
     uint256 GetHash() const;
 
     std::string ToString() const;
+
+    bool IsSpecialTxVersion() const
```
Comment (knst) [inline src/primitives/transaction.h:318]: let's drop it from CMutableTransaction which is **_struct_** not **_class_**.
Let's keep it simple as possible.

Check [this,](https://github.com/knst/dash/commit/986ce23fd9569d205aaabb017250243724f80cd0) [this](https://github.com/knst/dash/commit/0d3758a74af2ff28a0efd076ae69a871cbb3db3b) and [finally this](https://github.com/knst/dash/commit/f9c04d62d806fea5b0035010f6858d54da937d6c)

**Review by knst — CHANGES_REQUESTED** [review]

File: src/primitives/transaction.h (hunk 1 above)
Comment (UdjinM6) [reply src/primitives/transaction.h:318]: It's ok-ish to have small simple helpers in struct imo. Bitcoin even has a slightly complicated one https://github.com/bitcoin/bitcoin/blob/master/src/primitives/transaction.h#L412-L420. I also like 2 initial patches you provided btw, pls submit them in a separate PR if we agree on having helpers here.

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -1476,8 +1476,8 @@ class CTransactionSignatureSerializer
              SerializeOutput(s, nOutput);
         // Serialize nLockTime
         ::Serialize(s, txTo.nLockTime);
-        if (txTo.nVersion == 3 && txTo.nType != TRANSACTION_NORMAL)
-            ::Serialize(s, txTo.vExtraPayload);
+        if (txTo.HasExtraPayloadField())
+             ::Serialize(s, txTo.vExtraPayload);
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/script/interpreter.cpp:1480]: ... and also wrong (5 whitespaces instead of 4)
```suggestion
            ::Serialize(s, txTo.vExtraPayload);
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/primitives/transaction.h (hunk 1 above)
Comment (knst) [reply src/primitives/transaction.h:318]: but these helpers in CMutableTransaction are not actually useful, are they?

Comment (UdjinM6) [reply src/primitives/transaction.h:318]: that's true for `IsSpecialTxVersion` probably cause it's used once and is used internally only but `HasExtraPayloadField` avoids hardcoding tx version in `src/script/interpreter.cpp` which looks valuable

Comment (knst) [reply src/primitives/transaction.h:318]: `src/script/interpreter.cpp` use low-level own implementtion of serialization.

just look to the couple lines above:
```
        int32_t n32bitVersion = txTo.nVersion | (txTo.nType << 16);
```

it knows too much, that's still internal usage IMO even if it is outside of struct

Comment (UdjinM6) [reply src/primitives/transaction.h:318]: Hmm, true... ok, you convinced me I guess :D

Comment (UdjinM6) [reply src/primitives/transaction.h:318]: makes sense, one small suggestion https://github.com/knst/dash/commit/b645a1d3d34bec08e16f576ae03eb871d22d4e50#r138441182

**Comment by knst** [comment]
> wallet_multiwallet.py --usecli                   | ✖ Failed  | 23 s

I tried to re-start CI; everytime this test fails, maybe there's some newly introduced issue? I will try to re-start one more time

**Comment by panleone** [comment]
> > wallet_multiwallet.py --usecli                   | ✖ Failed  | 23 s
> 
> I tried to re-start CI; everytime this test fails, maybe there's some newly introduced issue? I will try to re-start one more time
> 
> The last failure: https://gitlab.com/dashpay/dash/-/jobs/6135684689

I don't know why it fails, but I have seen it failing in every PR I opened (always in the same job linux64_tsan-test)

**Comment by UdjinM6** [comment]
CI falure is unrelated, should be fixed by #5867

**Comment by PastaPastaPasta** [comment]
Hi @panleone; do you intend for this PR to move forward? I would see this likely merged soon with review addressed

File: src/script/interpreter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (panleone) [reply src/script/interpreter.cpp:1480]: I should stop running `clang-format-diff.py` to format lol

**Comment by panleone** [comment]
> Hi @panleone; do you intend for this PR to move forward? I would see this likely merged soon with review addressed

Completely forgot about this PR... reviews tackled but I do not have the time to make the history clean so I left 4 extra commits where I applied knst's patches

**Review by knst — DISMISSED** [review]
LGTM, seems as all my suggestions are included

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Comment by PastaPastaPasta** [comment]
Hi @panleone; Sorry, but this PR needs to be rebased, it has conflicts with develop.

**Comment by panleone** [comment]
> Hi @panleone; Sorry, but this PR needs to be rebased, it has conflicts with develop.

should be good now

**Review by knst — APPROVED** [review]
utACK 8b6c96d20839ce35f484d762afb7778f62de2a33

**Pull request merged**